resources/nspawn.go

```diff
@@ -0,0 +1,306 @@
+			}
+
+		case <-Startup(startup):
+			cuid.SetConverged(false)
+			send = true
+
+		// process org.freedesktop.machine1 events for this resource's
+		// name
+		case event := <-buschan:
+			if event.Body[0] == obj.GetName() {
+				if event.Name == machineNew {
+					log.Printf("Machine Started: %#v",
+						event.Body[0])
+				}
+				if event.Name == machineRemoved {
+					log.Printf("Machine Stopped: %#v",
+						event.Body[0])
+				}
+			}
+			send = true
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			startup = true // startup finished
+			send = false
+			if exit, err := obj.DoSend(processChan, ""); exit || err != nil {
+				return err // we exit or bubble up a NACK
+			}
+		}
+	}
+}
+
+// CheckApply is run to check the state and, if apply is true, to apply the
+// necessary changes to reach the desired state. this is run before Watch and
+// again if watch finds a change occurring to the state
+func (obj *NspawnRes) CheckApply(apply bool) (checkok bool, err error) {
+	log.Printf("%v[%v]: CheckApply(%t)", obj.Kind(), obj.GetName(), apply)
+
+	if obj.isStateOK { // cache the state
+		return true, nil
+	}
+
+	// this resource depends on systemd ensure that it's running
+	if !systemdUtil.IsRunningSystemd() {
+		return false, errors.New("Systemd is not running.")
+	}
+
+	// connect to org.freedesktop.machine1.Manager
+	conn, err := machined.New()
+
+	if err != nil {
+		return false, fmt.Errorf("Failed to connect to dbus: %s", err)
+	}
+
+	// compare the current state with the desired state and perform the
+	// appropriate action
+	var exists = true
+	properties, err := conn.GetProperties(obj.GetName())
+	if err != nil {
+		if err, ok := err.(dbus.Error); ok && err.Name !=
+			"org.freedesktop.machine1.NoSuchMachine" {
+			return false, err
+		}
+		exists = false
+		// if we could not successfully get the properties because
+		// there's no such machine the machine is stopped
+		// error if we need the image ignore if we don't
+		_, err = conn.GetImage(obj.GetName())
+		if err != nil && obj.State != stopped {
+			return false, fmt.Errorf(
+				"No machine nor image named '%s'",
+				obj.GetName())
+		}
+	}
+
+	// if the machine doesn't exist and is supposed to be stopped or
+	// the state matches we're done
+	if !exists && obj.State == stopped || properties["State"] == obj.State {
+		return true, nil
+	}
+
+	if obj.State == running {
+		// start the machine using svc resource
+		var serviceName = fmt.Sprintf(nspawnServiceTmpl,
+			obj.GetName())
+		log.Printf("Start service: %s", serviceName)
+		obj.svc.Name = serviceName
+		obj.svc.State = running
+		return obj.svc.CheckApply(apply)
+
+	}
+	if obj.State == stopped {
+		// terminate the machine with
+		// org.freedesktop.machine1.Manager.KillMachine
+		log.Printf("Stopping nspawn machine: %s", obj.GetName())
+		if apply {
+			if err := conn.KillMachine(obj.GetName()); err != nil {
+				return false, err
+			}
+		}
+		return true, nil
```

Comment:
What situation gets us here? This means "no error and state was as we expected and I made no changes".